### ISSUE TITLE
Support for storage device synchronization 

### ISSUE THREAD

**Issue by arsnyder16** [nih-at/libzip#471]
**Description**
Option to force synchronization to the storage device. Since the only guarantee from the OS when writing files is that the OS has accepted the change it would be nice allow forcing synchronization to disk. This is beneficial potentially in cloud environments where you may be writing to a remote NFS this ensures the file is full synced. 

Source reference for various other support for a similar feature

- [sqllite](https://github.com/sqlite/sqlite/blob/2eb9605e814275665ff0a24cbf910f8ce09a8470/src/os_unix.c#L3613)
- [redis](https://github.com/redis/redis/blob/701f06657d20ec42f0cf78de0ac9d7197e44c00c/src/config.h#L105)
- [mysql](https://github.com/mysql/mysql-server/blob/61a3a1d8ef15512396b4c2af46e922a19bf2b174/mysys/my_sync.cc#L85)
- [golang](https://pkg.go.dev/os#File.Sync)
  - [darwin](https://cs.opensource.google/go/go/+/master:src/internal/poll/fd_fsync_darwin.go?q=Fsync&ss=go%2Fgo)
  - [posix](https://cs.opensource.google/go/go/+/master:src/internal/poll/fd_fsync_posix.go?q=Fsync&ss=go%2Fgo)
  - [win](https://cs.opensource.google/go/go/+/master:src/internal/poll/fd_fsync_windows.go?q=Fsync&ss=go%2Fgo)

**Solution**
zip_sync(zip_t *archive);

Various platforms have different methods 
```c
// posix
// with F_FULLFSYNC, MAC OS
    fcntl(fileno(file), F_FULLFSYNC, 0)
// without full sync 
    fsync(fileno(file))

// windows
    _commit(_fileno(file))
// or
   FlushFileBuffers(file_handle)
```

**Comment by 0-wiz-0** [comment]
What is your expectation when you would call this function?

To be clear - libzip only writes out a file on the file system in `zip_close()`, which closes the archive, so before that, syncs don't make much sense.

You could perhaps argue that we should call fsync during zip_close, but that would just slow down the `zip_close()` operation for no apparent benefit, since we only replace any previously existing zip archive with the new one when we've finished writing the new one.

So I'm unclear on the benefit. Please explain.

**Comment by arsnyder16** [comment]
Yes I believe this would need to built into zip_close. 

In situations where zip files are being written to an NFS and you want a guarantee that it's been fully written before continuing. Similar reasons why database system do in a distributed environment 

Another situation one machine is writing out the file and then kicking off distributed work to other remote machines using the NFS, before issuing the map reduce function to the distributed machines you want to guarantee it's been fully synchronized.

**Comment by 0-wiz-0** [comment]
You could just manually call a sync function on the zip archive after libzip is done?

**Comment by arsnyder16** [comment]
> You could just manually call a sync function on the zip archive after libzip is done?

Agreed. This is the workaround but that also causes the entire system to flush not just the single file descriptor.

Just proposing the idea

**Comment by arsnyder16** [comment]
No problem, thanks for considering
